hazelcast/src/main/java/com/hazelcast/config/RingbufferConfig.java

```diff
@@ -0,0 +1,322 @@
+     * @return the total number of backups: backupCount plus asyncBackupCount
+     */
+    public int getTotalBackupCount() {
+        return backupCount + asyncBackupCount;
+    }
+
+    /**
+     * Creates a readonly copy of this RingbufferConfig.
+     *
+     * @return the readonly copy.
+     */
+    public RingbufferConfig getAsReadOnly() {
+        return new RingbufferConfigReadonly(this);
+    }
+
+    /**
+     * Gets the retention in seconds.
+     *
+     * @return the retention in seconds. Returns 0 if there is no retention.
+     */
+    public int getRetentionSeconds() {
+        return retentionSeconds;
+    }
+
+    /**
+     * Sets the retention in seconds.
+     *
+     * Retention is the time the ringbuffer is going to retain items before deleting them.
+     *
+     * Retention can also be disabled by setting retentionSeconds to 0. It means that items won't get removed because they
+     * retire. They will only overwrite. This means that when retention is disabled, that after tail did a full loop in the ring
+     * that the size will always be equal to the capacity.
+     *
+     * @param retentionSeconds the retention period in seconds
+     * @return the updated RingbufferConfig
+     * @throws IllegalArgumentException if retentionSeconds smaller than 0.
+     */
+    public RingbufferConfig setRetentionSeconds(int retentionSeconds) {
+        this.retentionSeconds = checkNotNegative(retentionSeconds, "retention can't be smaller than 0");
+        return this;
+    }
+
+    /**
+     * Gets the InMemoryFormat.
+     *
+     * @return the InMemoryFormat.
+     */
+    public InMemoryFormat getInMemoryFormat() {
+        return inMemoryFormat;
+    }
+
+    /**
+     * Sets the InMemoryFormat.
+     *
+     * Setting the InMemoryFormat controls format of storing an item in the ringbuffer:
+     * <ol>
+     *     <li>{@link InMemoryFormat#OBJECT}: the item is stored in deserialized format (so a regular object)</li>
+     *     <li>{@link InMemoryFormat#BINARY}: the item is stored in serialized format (so a is binary blob) </li>
+     * </ol>
+     *
+     * The default is binary. The object InMemoryFormat is useful when:
+     * <ol>
+     *     <li>of the object stored in object format has a smaller footprint than in binary format</li>
+     *     <li>if there are readers using a filter. Since for every filter invocation, the object needs to be available in
+     *     object format.</li>
+     * </ol>
+     *
+     * @param inMemoryFormat the new in memory format.
+     * @return the updated Config.
+     * @throws NullPointerException if inMemoryFormat is null.
+     * @throws IllegalArgumentException if {@link InMemoryFormat#NATIVE} in memory format is selected.
+     */
+    public RingbufferConfig setInMemoryFormat(InMemoryFormat inMemoryFormat) {
+        checkNotNull(inMemoryFormat, "inMemoryFormat can't be null");
+        checkFalse(inMemoryFormat == NATIVE, "InMemoryFormat " + NATIVE + " is not support");
```

Comment:
It doesn't make sense to add test the the InvalidConfigurationTest.

The RingbufferConfig is validated by the RingbufferConfigTest. There is 100% coverage and all bad inputs are tested. See included source for more info

The InvalidConfigurationTest repeats the same thing. There is no value on duplicating the tests, since the XMLConfigBuilder relies on the RingbufferConfig setters to set the fields.
![screenshot from 2015-05-04 10 06 32](https://cloud.githubusercontent.com/assets/105243/7449774/7c9eace2-f245-11e4-9dd0-d32c450855c6.png)

```
package com.hazelcast.config;

import com.hazelcast.test.HazelcastParallelClassRunner;
import com.hazelcast.test.annotation.QuickTest;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;
import org.mockito.internal.matchers.Null;

import java.util.concurrent.TimeUnit;

import static com.hazelcast.config.RingbufferConfig.DEFAULT_ASYNC_BACKUP_COUNT;
import static com.hazelcast.config.RingbufferConfig.DEFAULT_CAPACITY;
import static com.hazelcast.config.RingbufferConfig.DEFAULT_SYNC_BACKUP_COUNT;
import static com.hazelcast.partition.InternalPartition.MAX_BACKUP_COUNT;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.fail;

@RunWith(HazelcastParallelClassRunner.class)
@Category(QuickTest.class)
public class RingbufferConfigTest {

    public static final String NAME = "someringbuffer";

    @Test
    public void testDefaultSetting() {
        RingbufferConfig config = new RingbufferConfig(NAME);
        assertEquals(NAME, config.getName());
        assertEquals(DEFAULT_SYNC_BACKUP_COUNT, config.getBackupCount());
        assertEquals(DEFAULT_ASYNC_BACKUP_COUNT, config.getAsyncBackupCount());
        assertEquals(DEFAULT_CAPACITY, config.getCapacity());
    }

    @Test
    public void testCloneConstructor() {
        RingbufferConfig original = new RingbufferConfig(NAME);
        original.setBackupCount(2).setAsyncBackupCount(1).setCapacity(10);

        RingbufferConfig clone = new RingbufferConfig(original);

        assertEquals(original.getName(), clone.getName());
        assertEquals(original.getBackupCount(), clone.getBackupCount());
        assertEquals(original.getAsyncBackupCount(), clone.getAsyncBackupCount());
        assertEquals(original.getCapacity(), clone.getCapacity());
    }

    @Test
    public void testCloneConstructorWithName() {
        RingbufferConfig original = new RingbufferConfig(NAME);
        original.setBackupCount(2).setAsyncBackupCount(1).setCapacity(10);

        RingbufferConfig clone = new RingbufferConfig("foobar", original);

        assertEquals("foobar", clone.getName());
        assertEquals(original.getBackupCount(), clone.getBackupCount());
        assertEquals(original.getAsyncBackupCount(), clone.getAsyncBackupCount());
        assertEquals(original.getCapacity(), clone.getCapacity());
    }

    // =================== set capacity ===========================

    @Test
    public void setCapacity() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setCapacity(1000);

        assertEquals(1000, config.getCapacity());
    }

    @Test(expected = IllegalArgumentException.class)
    public void setCapacity_whenTooSmall() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setCapacity(0);
    }

    // =================== set backups count ===========================

    @Test
    public void setBackupCount() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setBackupCount(4);

        assertEquals(4, config.getBackupCount());
    }

    @Test(expected = IllegalArgumentException.class)
    public void setBackupCount_whenTooSmall() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setBackupCount(-1);
    }

    @Test(expected = IllegalArgumentException.class)
    public void setBackupCount_whenTooLarge() {
        RingbufferConfig ringbufferConfig = new RingbufferConfig(NAME);

        ringbufferConfig.setBackupCount(MAX_BACKUP_COUNT + 1);
    }

    // =================== set async backup count ===========================

    @Test
    public void setAsyncBackupCount() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setAsyncBackupCount(4);

        assertEquals(4, config.getAsyncBackupCount());
    }

    @Test(expected = IllegalArgumentException.class)
    public void setAsyncBackupCount_whenTooSmall() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setAsyncBackupCount(-1);
    }

    @Test(expected = IllegalArgumentException.class)
    public void setAsyncBackupCount_whenTooLarge() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setAsyncBackupCount(MAX_BACKUP_COUNT + 1);
    }

    // ============= get total backup count ====================

    @Test
    public void getTotalBackupCount() {
        RingbufferConfig config = new RingbufferConfig(NAME);
        config.setAsyncBackupCount(2);
        config.setBackupCount(3);

        int result = config.getTotalBackupCount();

        assertEquals(5, result);
    }

    // ================== retention =================================


    @Test(expected = IllegalArgumentException.class)
    public void setRetentionSeconds_whenNegative() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setRetentionSeconds(-1);
    }

    @Test
    public void setRetentionSeconds() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        RingbufferConfig returned = config.setRetentionSeconds(10);

        assertSame(returned, config);
        assertEquals(10, config.getRetentionSeconds());
    }

    // ================== inmemoryformat =================================


    @Test(expected = NullPointerException.class)
    public void setInMemoryFormat_whenNull() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setInMemoryFormat(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void setInMemoryFormat_whenNative() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        config.setInMemoryFormat(InMemoryFormat.NATIVE);
    }

    @Test
    public void setInMemoryFormat() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        RingbufferConfig returned = config.setInMemoryFormat(InMemoryFormat.OBJECT);

        assertSame(config, returned);
        assertEquals(InMemoryFormat.OBJECT, config.getInMemoryFormat());
    }

    // ==================== toString ================================

    @Test
    public void test_toString() {
        RingbufferConfig config = new RingbufferConfig(NAME);

        String s = config.toString();

        assertEquals("RingbufferConfig{name='someringbuffer', capacity=100000, backupCount=1, " +
                "asyncBackupCount=0, retentionSeconds=30, inMemoryFormat=BINARY}", s);
    }

    // =================== getAsReadOnly ============================

    @Test
    public void getAsReadOnly() {
        RingbufferConfig original = new RingbufferConfig(NAME);
        original.setBackupCount(2).setAsyncBackupCount(1).setCapacity(10).setRetentionSeconds(400);

        RingbufferConfig readonly = original.getAsReadOnly();
        assertNotNull(readonly);

        assertEquals(original.getName(), readonly.getName());
        assertEquals(original.getBackupCount(), readonly.getBackupCount());
        assertEquals(original.getAsyncBackupCount(), readonly.getAsyncBackupCount());
        assertEquals(original.getCapacity(), readonly.getCapacity());
        assertEquals(original.getRetentionSeconds(), readonly.getRetentionSeconds());
        assertEquals(original.getInMemoryFormat(), readonly.getInMemoryFormat());

        try {
            readonly.setCapacity(10);
            fail();
        } catch (UnsupportedOperationException expected) {
        }

        try {
            readonly.setAsyncBackupCount(1);
            fail();
        } catch (UnsupportedOperationException expected) {
        }

        try {
            readonly.setBackupCount(1);
            fail();
        } catch (UnsupportedOperationException expected) {
        }

        try {
            readonly.setRetentionSeconds(1);
            fail();
        } catch (UnsupportedOperationException expected) {
        }

        try {
            readonly.setInMemoryFormat(InMemoryFormat.OBJECT);
            fail();
        } catch (UnsupportedOperationException expected) {
        }
    }
}
```